src/Npgsql/Internal/TypeHandlers/ArrayHandler.cs

```diff
@@ -496,15 +494,28 @@ public ArrayHandlerWithPsv(PostgresType arrayPostgresType, NpgsqlTypeHandler ele
 
         protected internal override async ValueTask<TRequestedArray> ReadCustom<TRequestedArray>(NpgsqlReadBuffer buf, int len, bool async, FieldDescription? fieldDescription = null)
         {
-            if (ArrayTypeInfo<TRequestedArray>.ElementType == typeof(TElementPsv))
+            if (ArrayTypeInfo<TRequestedArray>.IsArray)
             {
-                if (ArrayTypeInfo<TRequestedArray>.IsArray)
-                    return (TRequestedArray)(object)await ReadArray<TElementPsv>(buf, async, typeof(TRequestedArray).GetArrayRank());
+                if (ArrayTypeInfo<TRequestedArray>.ElementType == typeof(TElementPsv))
+                    return (TRequestedArray)await ReadArray<TElementPsv>(buf, async, typeof(TRequestedArray).GetArrayRank());
 
-                if (ArrayTypeInfo<TRequestedArray>.IsList)
-                    return (TRequestedArray)(object)await ReadList<TElementPsv>(buf, async);
+                return (TRequestedArray)await ArrayTypeInfo<TRequestedArray>.ReadArray(this, buf, async);
             }
-            return await base.ReadCustom<TRequestedArray>(buf, len, async, fieldDescription);
+
+            // We evaluate List last to better support reflection free mode
```

Comment:
Curious - how does the ordering here affect reflection-free mode?

Reply:
Reflection free mode supports most of the array Type calls we do (GetElementType, GetRank etc) but it doesn't for any other types. Making sure we only call into ListInfo after that allows most code to work if it only cares about arrays.

Reply:
Thanks, makes sense.